speechbrain/lobes/models/wav2vec.py

```diff
@@ -0,0 +1,610 @@
+            Negative embeddings from feature extractor with shape (N, B, T, C)
+        where N is number of negatives.
+        """
+        neg_is_pos = (y == negs).all(-1)
+        y = y.unsqueeze(0)
+        targets = torch.cat([y, negs], dim=0)
+        logits = torch.cosine_similarity(
+            x.float(), targets.float(), dim=-1
+        ).type_as(x)
+
+        if neg_is_pos.any():
+            logits[1:][neg_is_pos] = float("-inf")
+        logits = logits.transpose(0, 2).reshape(-1, logits.size(0))
+
+        targets = torch.zeros(
+            (logits.size(0)), dtype=torch.long, device=logits.device
+        )
+        loss = F.cross_entropy(
+            logits / self.logit_temp, targets, reduction="sum"
+        )
+        accuracy = torch.sum(logits.argmax(-1) == 0) / (
+            logits.numel() / logits.size(-1)
+        )
+        return loss, accuracy
+
+
+def make_padding_mask(inp_shape, abs_lens):
+    """False when to be used, following speechbrain convention """
+    padding_mask = (
+        torch.arange(inp_shape[1], device=abs_lens.device)[None, :] >= abs_lens[:, None]
+    )
+    return padding_mask
+
+
+def compute_mask(shape, padding_mask, mask_prob, mask_length):
+    bs, padded_sample_len = shape
+    if padding_mask is not None:
+        sample_lens = padded_sample_len - torch.sum(padding_mask, -1)
+    else:
+        sample_lens = torch.ones(bs, dtype=torch.int32) * padded_sample_len
+
+    min_sample_len = sample_lens.min()
+    # So we dont have ragged tensors number of masks is the same for each sample.
+    num_mask = int(
+        mask_prob * min_sample_len / float(mask_length) + random.random()
+    )
+
+    mask_idcs = []
+    for i in range(bs):
+        sample_len = sample_lens[i].item()
+        mask_indices = np.random.choice(
+            sample_len - mask_length, num_mask, replace=False
+        )
+
+        mask_indices = np.asarray(
+            [
+                mask_indices[j] + offset
+                for j in range(len(mask_indices))
+                for offset in range(mask_length)
+            ]
+        )
+        mask_idcs.append(np.unique(mask_indices[mask_indices < sample_len]))
+
+    mask = np.full((bs, padded_sample_len), False)
+    # unique may have caused num masks to go down..
+    num_mask_total = num_mask * mask_length
+    for i, mask_idc in enumerate(mask_idcs):
+        # ..so we dont have ragged tensors we need to make the number of masked elements the same for each sample in the batch
+        if padding_mask is not None and len(mask_idc) < num_mask_total:
+            num_mask_missing = num_mask_total - len(mask_idc)
+            arange = np.arange(sample_lens[i].item())
+            arange = np.delete(arange, mask_idc)
+            extra_indcs = np.random.choice(
+                arange, num_mask_missing, replace=False
+            )
+            mask[i, extra_indcs] = True
+        mask[i, mask_idc] = True
+    return mask
+
+
+def apply_mask(
```

Comment:
The entire masking pipeline is hard to follow, we will need lot of documentation